### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by YegorPopinako**
implemented classes and base CRUD operations with DB

### CODE REVIEW

File: src/main/java/ua/petproject/model/Element.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package ua.petproject.model;
+import jakarta.persistence.Entity;
```
Comment (malyshevhen) [inline src/main/java/ua/petproject/model/Element.java:4]:
```suggestion
import static jakarta.persistence.GenerationType.IDENTITY;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;
import ua.petproject.categories.Category;

@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@EqualsAndHashCode(of = "id")
@ToString
@Entity
@Table(name = "elements")
public class Element {

    @Id
    @GeneratedValue(strategy = IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String name;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private Category category;
}

```

File: src/main/resources/application.properties [hunk 2]
```diff
@@ -1 +1,16 @@
...
+spring.datasource.url=jdbc:postgresql://localhost:5432/postgres
+spring.datasource.username=postgres
+spring.datasource.password=Qq245321
+spring.datasource.driver-class-name=org.postgresql.Driver
 
+# Enable Flyway
+spring.flyway.enabled=true
+
+# Specify locations of SQL migrations
+spring.flyway.locations=classpath:db/migration
```
Comment (malyshevhen) [inline src/main/resources/application.properties]: Delete default value

File: src/main/resources/application.properties [hunk 3]
```diff
@@ -1 +1,16 @@
+# Database Configuration
+spring.datasource.url=jdbc:postgresql://localhost:5432/postgres
+spring.datasource.username=postgres
+spring.datasource.password=Qq245321
+spring.datasource.driver-class-name=org.postgresql.Driver
 
+# Enable Flyway
+spring.flyway.enabled=true
```
Comment (malyshevhen) [inline src/main/resources/application.properties]: Turn off

File: src/main/java/ua/petproject/controller/CategoryController.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package ua.petproject.controller;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Controller;
+import ua.petproject.categoryservice.CategoryService;
+import ua.petproject.categoryservice.CategoryServiceIMPL;
```
Comment (malyshevhen) [inline src/main/java/ua/petproject/controller/CategoryController.java]: Delete unused import

File: src/main/java/ua/petproject/controller/CategoryController.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+import ua.petproject.model.Element;
+
+@Controller
+public class CategoryController {
+
+    @Autowired
```
Comment (malyshevhen) [inline src/main/java/ua/petproject/controller/CategoryController.java]: Use constructor injection instead

File: src/main/java/ua/petproject/controller/CategoryController.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
+package ua.petproject.controller;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Controller;
+import ua.petproject.categoryservice.CategoryService;
+import ua.petproject.categoryservice.CategoryServiceIMPL;
+import ua.petproject.model.Element;
+
+@Controller
```
Comment (malyshevhen) [inline src/main/java/ua/petproject/controller/CategoryController.java]:
```suggestion
@RestController
@RequestMapping("/api/categories")
```

File: src/main/java/ua/petproject/controller/CategoryController.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+public class CategoryController {
+
```
Comment (malyshevhen) [inline src/main/java/ua/petproject/controller/CategoryController.java]: // URL: /api/categories     POST GET(all)
    // URL: /api/categories/id  GET(byId) PUT(byId) DELETE(byId)

File: src/main/java/ua/petproject/controller/CategoryController.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    @Autowired
+    private CategoryService categoryService;
+
+    public Element getElement(long id) {
```
Comment (malyshevhen) [inline src/main/java/ua/petproject/controller/CategoryController.java]:
```suggestion
    public Element getElement(Long id) {
```

File: src/main/java/ua/petproject/categoryservice/CategoryService.java [hunk 9]
```diff
@@ -0,0 +1,11 @@
+package ua.petproject.categoryservice;
+
+import ua.petproject.model.Element;
+
+public interface CategoryService {
+    Element getElement(long id);
+    void deleteElement(long id);
+    Element addElement(Element element);
+    Element updateElement(long id, Element element);
+
```
Comment (malyshevhen) [inline src/main/java/ua/petproject/categoryservice/CategoryService.java]: Add getAll method

File: src/main/java/ua/petproject/categoryservice/CategoryService.java [hunk 10]
```diff
@@ -0,0 +1,11 @@
+package ua.petproject.categoryservice;
+
+import ua.petproject.model.Element;
+
+public interface CategoryService {
```
Comment (malyshevhen) [inline src/main/java/ua/petproject/categoryservice/CategoryService.java]:
```suggestion
public interface ElementService {
```

File: src/main/java/ua/petproject/categoryservice/CategoryService.java [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+    Element getElement(long id);
```
Comment (malyshevhen) [inline src/main/java/ua/petproject/categoryservice/CategoryService.java]: Add empty lines above each method

File: src/main/java/ua/petproject/categoryservice/CategoryService.java (hunk 9 above)
Comment (malyshevhen) [inline src/main/java/ua/petproject/categoryservice/CategoryService.java]: Add get all method

File: src/main/java/ua/petproject/categoryservice/CategoryServiceIMPL.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
+package ua.petproject.categoryservice;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import ua.petproject.model.Element;
+import ua.petproject.repository.CategoryDAO;
+
+@Service
+public class CategoryServiceIMPL implements CategoryService {
+
+    @Autowired
```
Comment (malyshevhen) [inline src/main/java/ua/petproject/categoryservice/CategoryServiceIMPL.java]: Use controller injection

File: src/main/java/ua/petproject/categoryservice/CategoryServiceIMPL.java [hunk 13]
```diff
@@ -0,0 +1,36 @@
+package ua.petproject.categoryservice;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import ua.petproject.model.Element;
+import ua.petproject.repository.CategoryDAO;
+
+@Service
+public class CategoryServiceIMPL implements CategoryService {
```
Comment (malyshevhen) [inline src/main/java/ua/petproject/categoryservice/CategoryServiceIMPL.java]:
```suggestion
public class CategoryServiceImpl implements CategoryService {
```
Rename class following the naming convention

File: src/main/java/ua/petproject/categoryservice/CategoryServiceIMPL.java [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+    @Autowired
+    private CategoryDAO categoryDAO;
+
+    @Override
+    public Element getElement(long id) {
+        return categoryDAO.findById(id).orElse(null);
```
Comment (malyshevhen) [inline src/main/java/ua/petproject/categoryservice/CategoryServiceIMPL.java]:
```suggestion
        return categoryDAO.findById(id).orElseThrow(() -> new NoSuchElementException("SOME MESSAGE"));
```

File: src/main/java/ua/petproject/categoryservice/CategoryServiceIMPL.java [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    @Override
+    public void deleteElement(long id) {
```
Comment (malyshevhen) [inline src/main/java/ua/petproject/categoryservice/CategoryServiceIMPL.java]: Check if element exists

File: src/main/java/ua/petproject/categoryservice/CategoryServiceIMPL.java [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+        return categoryDAO.save(element);
+    }
+
+    @Override
+    public Element updateElement(long id, Element element) {
+        if (categoryDAO.existsById(id)) {
+            element.setId(id);
+            return categoryDAO.save(element);
+        }
+        return null;
```
Comment (malyshevhen) [inline src/main/java/ua/petproject/categoryservice/CategoryServiceIMPL.java]:
```suggestion
        Element exitingElement = getById(id);
        // reset object fields
        return existingElement;
```

File: src/main/java/ua/petproject/repository/CategoryDAO.java [hunk 17]
```diff
@@ -0,0 +1,6 @@
+package ua.petproject.repository;
+import org.springframework.data.jpa.repository.JpaRepository;
+import ua.petproject.model.Element;
+
+public interface CategoryDAO extends JpaRepository<Element, Long> {
```
Comment (malyshevhen) [inline src/main/java/ua/petproject/repository/CategoryDAO.java]:
```suggestion
public interface ElementRepository extends JpaRepository<Element, Long> {
```

File: src/main/java/ua/petproject/repository/CategoryDAO.java [hunk 18]
```diff
@@ -0,0 +1,6 @@
+package ua.petproject.repository;
+import org.springframework.data.jpa.repository.JpaRepository;
```
Comment (malyshevhen) [inline src/main/java/ua/petproject/repository/CategoryDAO.java]: Add new line

**Review by malyshevhen — CHANGES_REQUESTED** [review]
Fix the code.

**Review by Anatolii-Omelchenko — CHANGES_REQUESTED** [review]
Make corrections to comments.

**Pull request merged**